internal/transformer/kafka/producer.go

```diff
@@ -2,55 +2,66 @@ package kafka
 
 import (
 	"log"
+	"time"
 
+	_instrument "github.com/etf1/kafka-transformer/internal/instrument"
+	"github.com/etf1/kafka-transformer/pkg/instrument"
 	"github.com/etf1/kafka-transformer/pkg/logger"
 	confluent "gopkg.in/confluentinc/confluent-kafka-go.v1/kafka"
 )
 
 // Producer represents the kafka producer which will produce
 // the transformed message to a topic defined in the Message
 type Producer struct {
-	producer *confluent.Producer
-	log      logger.Log
+	producer  *confluent.Producer
+	log       logger.Log
+	collector instrument.Collector
 }
 
 // NewProducer constructor for Producer
-func NewProducer(l logger.Log, config *confluent.ConfigMap) (Producer, error) {
-	p, err := confluent.NewProducer(config)
+func NewProducer(l logger.Log, config *confluent.ConfigMap, collector instrument.Collector) (Producer, error) {
+	kafkaProducer, err := confluent.NewProducer(config)
 
 	if err != nil {
 		return Producer{}, err
 	}
 
+	p := Producer{
+		producer:  kafkaProducer,
+		log:       l,
+		collector: collector,
+	}
+
 	go func() {
 		defer log.Println("kafka producer stopped")
-		for e := range p.Events() {
+		for e := range kafkaProducer.Events() {
 			switch ev := e.(type) {
 			case *confluent.Message:
-				m := ev
-				if m.TopicPartition.Error != nil {
-					l.Errorf("delivery failed: %v", m.TopicPartition.Error)
+				msg := ev
+				if msg.TopicPartition.Error != nil {
+					p.collectAfter(msg, msg.TopicPartition.Error)
+					l.Errorf("delivery failed: %v", msg.TopicPartition.Error)
 				} else {
+					p.collectAfter(msg, nil)
 					l.Debugf("producer: delivered message to topic %s [%d] at offset %v",
-						*m.TopicPartition.Topic, m.TopicPartition.Partition, m.TopicPartition.Offset)
+						*msg.TopicPartition.Topic, msg.TopicPartition.Partition, msg.TopicPartition.Offset)
 				}
 			case confluent.Error:
+				p.collectAfter(nil, ev)
```

Comment:
@fkarakas This will raise an error because you are sending a `nil` value into `msg` argument and later, in the `collectAfter()` method, there is:
```go
th := msg.Opaque.(_instrument.TimeHolder)
```

Error raised:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x80 pc=0x57121a]
goroutine 24 [running]:
github.com/etf1/kafka-transformer/internal/transformer/kafka.Producer.collectAfter(0xc00007e380, 0xaeb320, 0xc000021040, 0xaeb2e0, 0xc0002cbdd0, 0x0, 0xae7e40, 0xc00020cb80)
        /project/vendor/github.com/etf1/kafka-transformer/internal/transformer/kafka/producer.go:86 +0x5a
github.com/etf1/kafka-transformer/internal/transformer/kafka.NewProducer.func1(0xc00007e380, 0xc00007e380, 0xaeb320, 0xc000021040, 0xaeb2e0, 0xc0002cbdd0, 0xaeb320, 0xc000021040)
        /project/vendor/github.com/etf1/kafka-transformer/internal/transformer/kafka/producer.go:50 +0x287
created by github.com/etf1/kafka-transformer/internal/transformer/kafka.NewProducer
        /project/vendor/github.com/etf1/kafka-transformer/internal/transformer/kafka/producer.go:35 +0x140
```

Could you check for `nil` values, maybe in `collectBefore()` and `collectAfter()` please? Or maybe you should never call these methods with a nil object?

Reply:
Missing a test here, when an error has occured.
So il will write the test in order to reproduce and fix it.
Thank you @eko